core/src/main/java/org/apache/iceberg/io/RollingFileWriter.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * A rolling writer capable of splitting incoming data/deletes into multiple files within one spec/partition
+ * based on the target file size.
+ */
+abstract class RollingFileWriter<T, W extends FileWriter<T, R>, R> implements FileWriter<T, R> {
+  private static final int ROWS_DIVISOR = 1000;
+
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final long targetFileSizeInBytes;
+  private final PartitionSpec spec;
+  private final StructLike partition;
+
+  private EncryptedOutputFile currentFile = null;
+  private long currentFileRows = 0;
+  private W currentWriter = null;
+
+  private boolean closed = false;
+
+  protected RollingFileWriter(OutputFileFactory fileFactory, FileIO io, long targetFileSizeInBytes,
+                              PartitionSpec spec, StructLike partition) {
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.targetFileSizeInBytes = targetFileSizeInBytes;
+    this.spec = spec;
+    this.partition = partition;
+  }
+
+  protected abstract W newWriter(EncryptedOutputFile file);
+
+  protected abstract void addResult(R result);
+
+  protected abstract R aggregatedResult();
+
+  protected PartitionSpec spec() {
+    return spec;
+  }
+
+  protected StructLike partition() {
+    return partition;
+  }
+
+  public CharSequence currentFilePath() {
```

Comment:
I removed the precondition here that the current file is not null. We will call this method for every single row while writing CDC records. Right now, `currentFile` is never null as we init it in the constructor.

Reply:
Sounds good to me. You mean the CDC writer constructor?

Reply:
I mean all classes that extend `RollingFileWriter` init the writer immediately so we shouldn't worry about the current file being null.

Reply:
Yeah, I saw that. I don't have a good way around this, but at least we know that it will fail quickly without that init call.